### ISSUE TITLE
Suggested addition to help URL consistency for Share/Like and SEO

### ISSUE THREAD

**Issue by toby07801** [JoomGallery/JoomGallery#23]
I've been using JoomGallery for a few years now and one of the problems I found with it was the fact that images and categories end up being accessible by so many different URLs which means search engines index duplicates and shares and likes on social media are not aggregated. As long as the last part of the link matches a valid image it shows up in the browser as a valid 200. I.e. 

www.domain.com/gallery/cat-1/subcat-1/image-id1
www.domain.com/gallery/image-id1
www.domain.com/gallery/cat-1/cat-1/cat-1/image-id1
www.domain.com/gallery/cat-5/image-id1

all display the same image under different URLs.

I've managed to fix this unwanted behaviour by adding a canonical url tag to my pages - I've written a bit of code which traverses all parent categories of an image and adds these to the "preferred" version of the url.

For example, the code below does this for a detail image view and I've added this to com_joomgallery\views\detail\view.html.php which seems to work fine.

    // Add Canonical URL to Detail Image 
    $cpath  = null;
    foreach($parents as $parent)
    {
      // Get a db connection.
      $db = JFactory::getDbo();
      // Create a new query object 
      $query = "SELECT alias"
      . " FROM #__joomgallery_catg"
      . " WHERE cid = $parent->cid"
      . " LIMIT 1"
      ;
      $db->setQuery($query);
      $rows = $db->loadObjectList();
      $itemrow = $rows[0];
      $cpath  .= $itemrow->alias.'/';
    }
    
    $cpath .= $image->alias;
     
    $canonicalurl = 'http://yourdomainname.com/gallery/'.$cpath;
    $this->_doc->addCustomTag('<link rel="canonical" href="'.$canonicalurl.'" />');
    $this->_doc->addCustomTag('<meta property="og:url" content="'.$canonicalurl.'" />');
    $this->_doc->addCustomTag('<meta name="twitter:url" content="'.$canonicalurl.'" />');
    $this->assignRef( 'canonicalurl', $canonicalurl);

Once this is added facebook/twitter/google etc will start using a consistent url for all images which is composed of all its parent categories in the correct order, no matter how messed up the link was to begin with so shares and likes will get aggregated and google will only index one version of the page.

Perhaps this is something that could be implemented in future versions for the benefit of everybody else?

PS: Forgive me if this is not the best PHP in the world I'm not really much of an advanced coder ;)

**Comment by Chraneco** [comment]
This seems like a good idea!
I have only one doubt. Are you already using this on a live site? If yes, is the system SEF plugin enabled on that site? If the parameter of that plugin is filled, it also adds a ```canonical``` link to the header. This way we would end up with two of these links. Not sure what Google will do with that in this case.

**Comment by toby07801** [comment]
I've been using this for about a year and a half without any issues. You can see it in action here: http://bit.ly/1PRWqyX (Have a look at the page source and search for "canonical")

The SEF plugin is enabled on this site and I have the domain added as a parameter but I just double-checked and I seem to have only the manually added canonicals added to the page headers of JoomGallery pages as per the above code, no canonicals added on any non-JoomGallery pages and whether I have entered a domain in the parameters of SEF or not doesn't seem to make any difference either. 

It might well need a bit of tweaking to be able to configure this through the JoomGallery config but the general idea is there I think. Maybe it's useful and worth implementing.

Here's the code I used to do the same for Categories (no idea with my limited php skills to know if this is the best and most efficient way to achieve this behaviour but it seems to work ok for me)

Added to **com_joomgallery/views/category/view.html.php**

     // Add canonical URL to Category Pages
     $cpath  = null;
     $calias = null;
	  // Get a db connection.
		$db = JFactory::getDbo();
	 
		// Create a new query object 
		$query = "SELECT alias"
		. " FROM #__joomgallery_catg"
		. " WHERE cid = $cat->cid"
		. " LIMIT 1"
		;
		$db->setQuery($query);
		$rows = $db->loadObjectList();
		$itemrow = $rows[0];
		
        $calias  .= $itemrow->alias;

      foreach($parents as $parent)
      {
		// Get a db connection.
		$db = JFactory::getDbo();
	 
		// Create a new query object 
		$query = "SELECT alias"
		. " FROM #__joomgallery_catg"
		. " WHERE cid = $parent->cid"
		. " LIMIT 1"
		;
		$db->setQuery($query);
		$rows = $db->loadObjectList();
		$itemrow = $rows[0];
		
        $cpath  .= $itemrow->alias.'/';
      }

      $cpath .= $calias;
	  
	  $canonicalurl = 'http://yourdomainname.com/gallery/'.$cpath;
	  $this->_doc->addCustomTag('<link rel="canonical" href="'.$canonicalurl.'" />');
	  $this->_doc->addCustomTag('<meta property="og:url" content="'.$canonicalurl.'" />');
	  $this->_doc->addCustomTag('<meta name="twitter:url" content="'.$canonicalurl.'" />');
	  $this->assignRef('canonicalurl', $canonicalurl);

**Comment by Chraneco** [comment]
Thanks for the link!
For example if you have a look at the source code of page ```/gallery/portraiture```, there are two canonical links in the header: One right at the top and one further down in the HTML.

We have to find a way to prevent that if we want to add this as a new feature to the gallery.

**Comment by toby07801** [comment]
I know I was wrong, just updated my earlier comment to correct myself but you were quicker, sorry! 
However, I've just checked some of my links on http://www.seoreviewtools.com/ and even though there are two canonicals it seems to pick the right one each time... Maybe the latest simply takes precedence? Worth checking with google how they handle this.

**Comment by toby07801** [comment]
Google says if there's more than one, all will be ignored. So yes, we'd have to find a way to pull Joomla's canonicals from JoomGallery pages if enabled and replace it with our own version... the Joomla version is not working out the correct link in the same way and not adding all the parents into the path unfortunately. :(

**Comment by toby07801** [comment]
This canonical thing must be new to Joomla 3 because I swear I didn't have this issue with Joomla 2.5 when I tested it. (I only upgraded a couple of months ago) 

I've just been experimenting with this SEF plugin settings and it looks like if I fill the parameter of the Joomla SEF plugin with my site's domain name then it does NOT put an additional canonical into the head. If I leave it blank (as it was by default) it does put a canonical in.

**Comment by Chraneco** [comment]
Yeah, there should be at least a way to disable that functionality in the SEF plugin because right now it puts that link even if the input field is left empty.

**Comment by toby07801** [comment]
Agreed, especially if it doesn't actually provide a unique canonical as can be demonstrated in the case of JoomGallery by taking out a couple of parent categories in the URL and observing the canonical tag added by Joomla change each time, when really it should always stay the same. 

As for my site, I've been happily using the above tweak for about two years now, but since the upgrade to Joomla 3 I have had to resort to disabling the canonical function in the SEF plugin to avoid the problem of two different canonical tags on the same page.
